src/perception/zed_wrapper/zed_wrapper.cpp

```diff
@@ -0,0 +1,258 @@
+            mPnh.param("svo_file", svoFile, {});
+            mPnh.param("ues_builtin_visual_odom", mUseBuiltinPosTracking, false);
+
+            if (imageWidth < 0 || imageHeight < 0) {
+                throw std::invalid_argument("Invalid image dimensions");
+            }
+            if (mGrabTargetFps < 0) {
+                throw std::invalid_argument("Invalid grab target framerate");
+            }
+
+            sl::InitParameters initParameters;
+            if (!svoFile.empty()) {
+                initParameters.input.setFromSVOFile(svoFile.c_str());
+            }
+            initParameters.camera_resolution = static_cast<sl::RESOLUTION>(resolution);
+            initParameters.depth_mode = static_cast<sl::DEPTH_MODE>(depthMode);
+            initParameters.coordinate_units = sl::UNIT::METER;
+            initParameters.sdk_verbose = true; // Log useful information
+            initParameters.camera_fps = mGrabTargetFps;
+            initParameters.coordinate_system = sl::COORDINATE_SYSTEM::RIGHT_HANDED_Z_UP_X_FWD; // Match ROS
+
+            if (mZed.open(initParameters) != sl::ERROR_CODE::SUCCESS) {
+                throw std::runtime_error("ZED failed to open");
+            }
+
+            sl::PositionalTrackingParameters positionalTrackingParameters;
+            mZed.enablePositionalTracking(positionalTrackingParameters);
+
+            mGrabThread = std::thread(&ZedNodelet::grabUpdate, this);
+            mPcThread = std::thread(&ZedNodelet::pointCloudUpdate, this);
+
+        } catch (std::exception const& e) {
+            NODELET_FATAL("Exception while starting: %s", e.what());
+            ros::shutdown();
+        }
+    }
+
+    /**
+     * Only relevant if direct tag detection is true.
+     * This handles loading a tag detection nodelet and passing it point clouds from the grab thread.
+     */
+    void ZedNodelet::pointCloudUpdate() {
+        try {
+            NODELET_INFO("Starting point cloud thread");
+
+            if (mDirectTagDetection) {
+                // TODO: ugly, this prevents OpenCV fast alloc from crashing
+                std::this_thread::sleep_for(100ms);
+
+                mTagDetectorNode = boost::make_shared<TagDetectorNodelet>();
+                mTagDetectorNode->init("tag_detector", getRemappingArgs(), getMyArgv());
+
+                // TODO: figure out why removing this causes a segfault
+                cv::aruco::getPredefinedDictionary(cv::aruco::DICT_4X4_50);
+            }
+
+            while (ros::ok()) {
+                mPcThreadProfiler.reset();
+
+                {
+                    std::unique_lock lock{mGrabMutex};
+                    mGrabDone.wait(lock);
+                }
+                ros::Time grabTime{slTime2Ros(mZed.getTimestamp(sl::TIME_REFERENCE::IMAGE))};
+                mPcThreadProfiler.addEpoch("Wait");
+
+                if (mZed.retrieveImage(mLeftImageMat, sl::VIEW::LEFT, sl::MEM::CPU, mImageResolution) != sl::ERROR_CODE::SUCCESS)
```

Comment:
I take it you confirmed via profiling that retrieveImage is fast (and grab is the slow thing)?

Reply:
They are both slow. From my home computer with Quality in 720p:

```
[ INFO] [1680495207.967549245]: [2178970924159706529] Total: 17ms
[ INFO] [1680495207.967597341]: 	Sensor data: 0ms
[ INFO] [1680495207.967621506]: 	Positional tracking: 0ms
[ INFO] [1680495207.967644537]: 	Grab: 16ms
[ INFO] [1680495208.889353428]: [13210474076691361599] Total: 33ms
[ INFO] [1680495208.889507241]: 	Point cloud publish: 0ms
[ INFO] [1680495208.889695749]: 	Direct Tag Detection: 0ms
[ INFO] [1680495208.889770072]: 	Fill Message: 1ms
[ INFO] [1680495208.889833420]: 	Retrieve: 18ms
[ INFO] [1680495208.889920846]: 	Image + camera info publish: 0ms
[ INFO] [1680495208.890014916]: 	Wait: 12ms
```

Reply:
The thing is I want the grab thread to be as fast as possible since it handles positional tracking.